wagtail/core/collectors.py

```diff
@@ -0,0 +1,669 @@
+        in the ``StreamField`` of this collector, ``self.field``.
+
+        A path is a ``tuple`` containing all the ``Block`` instances leading to
+        a ``Block`` instance that is a leaf of the tree.
+
+        Example:
+
+        >>> from wagtail.core.blocks import CharBlock
+        >>> from wagtail.core.fields import StreamField
+        >>> stream_field = StreamField([
+        ...     ('title', CharBlock()),
+        ...     ('items', ListBlock(CharBlock())),
+        ... ])
+        >>> [[b.__class__.__name__ for b in path]
+        ...  for path in StreamFieldCollector(stream_field)
+        ...              .block_tree_paths(stream_field.stream_block)]
+        [['StreamBlock', 'CharBlock'], ['StreamBlock', 'ListBlock', 'CharBlock']]
+        """
+        if isinstance(block, (StreamBlock, StructBlock)):
+            for child_block in block.child_blocks.values():
+                yield from self.block_tree_paths(child_block,
+                                                 ancestors + (block,))
+        elif isinstance(block, ListBlock):
+            yield from self.block_tree_paths(block.child_block,
+                                             ancestors + (block,))
+        else:
+            yield ancestors + (block,)
+
+    def find_block_type(self, block_types):
+        """
+        Yields all the possible paths for ``Block`` instances of one of the
+        ``Block`` subclasses passed as the ``block_types``.
+
+        Example:
+
+        >>> from wagtail.core.blocks import CharBlock, TextBlock
+        >>> from wagtail.core.fields import StreamField
+        >>> stream_field = StreamField([
+        ...     ('title', CharBlock()),
+        ...     ('items', ListBlock(TextBlock())),
+        ... ])
+        >>> [[b.__class__.__name__ for b in path]
+        ...  for path in StreamFieldCollector(stream_field)
+        ...              .find_block_type((CharBlock, TextBlock))]
+        [['StreamBlock', 'CharBlock'], ['StreamBlock', 'ListBlock', 'TextBlock']]
+        """
+        for block_path in self.block_tree_paths(self.field.stream_block):
+            if isinstance(block_path[-1], block_types):
+                yield block_path
+
+    def find_objects(self, value, block_path):
+        """
+        Recursively traverse a StreamField value to find all sub-values corresponding to
+        the given block path, and yield all model instances found within those sub-values.
+
+        :param value: A value existing at any level of ``StreamField`` data. The type of
+            this value must match the first item in ``block_path``.
+        :param block_path: The path to a block, as a tuple of block definition objects.
+        """
+        current_block, *remaining_path = block_path
+
+        if not remaining_path:
+            # End of path reached; return model instances found within `value`
+            if isinstance(current_block, RichTextBlock) and value is not None:
```

Comment:
I doubt anyone will actually do it, but what if someone writes a block type returning a `RichText` object? It will not be handled by this line, while checking for the value will handle such cases. This is why I originally checked the data types instead of the block types.

Also, is it possible to end up with a `None` value in a `RichTextBlock`, apart from a manual database rewrite?

(Making this review in the context of 19952a6bd2e95efc63af9b450183813450b93d8e)

Reply:
Good point! We need to check against the block definitions for the non-leaf cases, so I was following the same pattern here for consistency... but checking the value type would indeed be more flexible. Will update...

I guess a `RichTextBlock` can't (currently) be `None`, but I think it's good to be cautious here.